compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -196,32 +196,67 @@ fn project_and_unify_type<'cx, 'tcx>(
     debug!(?obligation, "project_and_unify_type");
 
     let mut obligations = vec![];
-    let normalized_ty = match opt_normalize_projection_type(
```

Comment:
I think the right approach is not to duplicate here. After all, `projection_ty` is a `Term`, so it goes as input into `opt_normalize_projection_type`. I think that `opt_normalize_projection_type` should be returning a `Term` and then this function here should just keep working as it was. 

Afaict you'll need to patch `confirm_candidate` to return a `Term`, and then most of the changes you did in this file become changing `Ty` to `Term` instead of duplicating the logic

Reply:
We could always create a wrapper just for `normalize_projection_type` that does the `unwrap` internally to make sure that all the callers are unaffected. Most of them do seem to always return types, but I'm not sure what will happen when users start using associated const bounds and accidentally use a const bound when they wanted a type bound or vice versa.